unison-runtime/src/Unison/Runtime/Foreign/Function.hs

```diff
@@ -49,165 +163,1313 @@ import Unison.Type
+  readForeign ![] !_ = foreignCCError "Int"
+  {-# INLINE readForeign #-}
+  writeForeign !stk !i = do
     stk <- bump stk
     stk <$ pokeI stk i
+  {-# INLINE writeForeign #-}
 
 instance ForeignConvention Word64 where
-  readForeign (i : args) stk = (args,) <$> peekOffN stk i
-  readForeign [] _ = foreignCCError "Word64"
-  writeForeign stk n = do
+  readForeign !(i : args) !stk = (args,) <$> peekOffN stk i
+  readForeign ![] !_ = foreignCCError "Word64"
+  {-# INLINE readForeign #-}
+  writeForeign !stk !n = do
     stk <- bump stk
     stk <$ pokeN stk n
+  {-# INLINE writeForeign #-}
 
 -- We don't have a clear mapping from these types to Unison types, most are just mapped to Nats.
 
 instance ForeignConvention Word8 where
   readForeign = readForeignAs (fromIntegral :: Word64 -> Word8)
+  {-# INLINE readForeign #-}
   writeForeign = writeForeignAs (fromIntegral :: Word8 -> Word64)
+  {-# INLINE writeForeign #-}
 
 instance ForeignConvention Word16 where
   readForeign = readForeignAs (fromIntegral :: Word64 -> Word16)
+  {-# INLINE readForeign #-}
   writeForeign = writeForeignAs (fromIntegral :: Word16 -> Word64)
+  {-# INLINE writeForeign #-}
 
 instance ForeignConvention Word32 where
   readForeign = readForeignAs (fromIntegral :: Word64 -> Word32)
+  {-# INLINE readForeign #-}
   writeForeign = writeForeignAs (fromIntegral :: Word32 -> Word64)
+  {-# INLINE writeForeign #-}
 
 instance ForeignConvention Char where
-  readForeign (i : args) stk = (args,) <$> peekOffC stk i
-  readForeign [] _ = foreignCCError "Char"
-  writeForeign stk ch = do
+  readForeign !(i : args) !stk = (args,) <$> peekOffC stk i
+  readForeign ![] !_ = foreignCCError "Char"
+  {-# INLINE readForeign #-}
+  writeForeign !stk !ch = do
     stk <- bump stk
     stk <$ pokeC stk ch
+  {-# INLINE writeForeign #-}
 
 instance ForeignConvention Val where
-  readForeign (i : args) stk = (args,) <$> peekOff stk i
-  readForeign [] _ = foreignCCError "Val"
-  writeForeign stk v = do
+  readForeign !(i : args) !stk = (args,) <$> peekOff stk i
+  readForeign ![] !_ = foreignCCError "Val"
+  {-# INLINE readForeign #-}
+  writeForeign !stk !v = do
     stk <- bump stk
     stk <$ (poke stk =<< evaluate v)
+  {-# INLINE writeForeign #-}
 
 -- In reality this fixes the type to be 'RClosure', but allows us to defer
 -- the typechecker a bit and avoid a bunch of annoying type annotations.
 instance ForeignConvention Closure where
-  readForeign (i : args) stk = (args,) <$> bpeekOff stk i
-  readForeign [] _ = foreignCCError "Closure"
-  writeForeign stk c = do
+  readForeign !(i : args) !stk = (args,) <$> bpeekOff stk i
+  readForeign ![] !_ = foreignCCError "Closure"
+  {-# INLINE readForeign #-}
+  writeForeign !stk !c = do
     stk <- bump stk
     stk <$ (bpoke stk =<< evaluate c)
+  {-# INLINE writeForeign #-}
 
 instance ForeignConvention Text where
   readForeign = readForeignBuiltin
+  {-# INLINE readForeign #-}
```

Comment:
I think we definitely _do_ want GHC to inline this, since obviously it's small, and will mean GHC specializes it rather than leaving things polymorphic. 

So I'm assuming you're just saying it's redundant because GHC will almost certainly inline it, but in that case, do you have a specific rule for how complex definitions should be before I add an explicit inline?

I tried removing the annotations from all the `ForeignConvention` instances to see what would happen and that slowed down the `tvarForeigns` benchmark by 2ms; so GHC isn't choosing to inline all of them.

I can just remove it from all the `x = y` cases if you like, but personally I'd prefer in this case to leave the inlines on all of them since:

1. We _do_ want them to be inlined
2. It's likely to reduce the chances that future ForeignConvention instances accidentally miss an Inline
3. If it ever gets moved to a different module GHC probably _won't_ inline it without the pragma.

Just let me know what you'd prefer and I'll try that and benchmark it.

Reply:
I'm not saying we don't want them to be inlined. I'm saying `INLINE` pragmas on `x = y` definitions are just telling GHC to do things it will decide to do anyway. It's not true that it only does cross-module inlining if you tell it to. And as far as I can tell, type classes don't introduce an obstacle, either.